src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolderManager2.java

```diff
@@ -522,7 +522,7 @@ static int compareShellFolders(Win32ShellFolder2 sf1, Win32ShellFolder2 sf2) {
         boolean special1 = sf1.isSpecial();
         boolean special2 = sf2.isSpecial();
 
-        if (special1 || special2) {
+        if (special1 && special2) {
```

Comment:
Could you please say something on how/why this change solves it ?

Reply:
It is well described in a [JBS comment](https://bugs.openjdk.org/browse/JDK-8305072?focusedId=14610400&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14610400) by the submitter.

Let's consider the failing case where:
```text
a = C:\Users\<user>\Documents(true)
b = C:\Users\<user>(false)
c = C:\Users\<user>\Documents(false)
```
The value in parentheses is `isSpecial()` flag.

When `a` and `c` are compared, `special1 = true` and `special2 = false`, therefore the code flow enters this `if`-block. Because the `equals` method doesn't take into account the `isSpecial` flag, both `i1` and `i2` are 0.

https://github.com/openjdk/jdk/blob/7bccc847fe9e8c5e640d8fa9c93ce32563a0ccfa/src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolderManager2.java#L534-L535

Thus, `a` and `c` are considered equal, which breaks the general contract of [`Comparable.compareTo`](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/Comparable.html#compareTo(T)).

If you look below this `if`-statement:

https://github.com/openjdk/jdk/blob/7bccc847fe9e8c5e640d8fa9c93ce32563a0ccfa/src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolderManager2.java#L545-L550

The comment and the code say that any special folder sorts above any regular folder. Therefore the code shouldn't enter that `if`-block above if only one of the folders is *special* — the order is already well-defined for such case.

The purpose of the `if`-block is to make some of the special folders even more special. In other words, it ensures the following special folders—`getPersonal`, `getDesktop`, `getDrives`, `getNetwork`—are at the top of the list, above any other special folders.

Reply:
@uckelman, Your message is hidden until you accept OpenJDK Terms of Use.

Reply:
Thank you, @uckelman, for the detailed analysis. Greatly appreciated.

Reply:
@uckelman Is there a way to reproduce the original bug in the user's environment?

I think there's a bug in JDK which leads to two _Documents_ folders in the list. If the bug is still reproducible, collecting more data will help us identify the problem and fix it.

Reply:
If I can get in contact with the user, I expect he'd still be able to trigger the bug with my test case, yeah. I'll try to get his attention.

Reply:
I haven't heard back yet. I hope investigating the second bug will not hold up the fix for the first bug getting merged.

Reply:
> I haven't heard back yet. I hope investigating the second bug will not hold up the fix for the first bug getting merged.

It's not a show-stopper. These are different bugs. If possible, I'd like to follow up and find the root cause of why there are two identical folders in the file list.